### DISCUSSION TITLE
Dependency Injection

### DISCUSSION THREAD

**Discussion by harperjohn** [BlazingSaddlesSolutions/TimeKeeper#21] [General]
When I first started looking into Blazor, I looked into Autofac because I thought that I had to use an external library... Then I found out that MS DI had matured and (apparently) was quite usable... so I put Autofac on the back burner, and mostly out of my mind. 

But then the "3 scopes" options of MS DI started to feel a bit limiting... especially since "scooped" seems to default/degenerate to "singleton" for Blazor server apps...

I feel the the MS DI is not "keeping up" with Blazor or maybe it is just too simple/rigid in general.

I could go either way MS DI or Autofac DI... but in either case, I think we should really understand/define our use of DI

I think DI is very useful and effective, but I feel that most people use it in a "boilerplate" / "automatic-reflex" way which can lessen (if not completely undermine) its use.

**Comment by AnkurSheel** [comment]
My 2 cents - I am happy to use MS DI especially as it's a learning project. 
It's also easy to migrate from MS DI to Autofac etc if you start hitting running into problems at scale. You don't hit problems with Singletons in DI till you start serving a large volume of requests. 

Also,  avoids the need to learn another library

**Comment by DotNetDublin** [comment]
I would agree with that.

When the application is mature we could look at more advanced topics and evaluate how they could benefit.

**Comment by jonhilt** [comment]
Yep, definitely in favour of keeping things simple to start.

Migrating is pretty straightforward these days, usually only needing
some changes to startup.cs/program.cs so using MS DI now shouldn't block
us from considering alternatives down the line.

**Comment by harperjohn** [comment]
Yes - I am definitely fine with default MS DI.

The main reason that I personally started to reconsider Autofac was that I was watching a video from one of my courses and the instructor added and used Autofac so quickly and easily...

It made me wonder. For example Blazor does not  really have a (full) set of GUI components, so using a third party library (or rolling your own) is a given.  I wondered if the MS DI was a "placeholder" that we were expected to fill with a "real" DI system. 

So then my view of Autofac kind of shifted from "big/complex/overkill" to well thought out/ well documented/ flexible / best-of-breed.

Again, I don't have a lot of (actual/real) experience with either MS DI or Autofac, but it occurred to me that others may have this level of experience so I figured it was worth "putting on the table".

It seems like MS DI is "good enough" and "easily replaceable" so MS DI seems fine.

**Comment by jonhilt** [comment]
I think the main reason I wouldn't really mind what we start with is primarily because it has no impact on how we end up building our components.

In all cases, whether using MS DI, Autofac or anything else, we'll still be using the [standard `@inject` Razor directive](https://docs.microsoft.com/en-us/aspnet/core/blazor/fundamentals/dependency-injection?view=aspnetcore-5.0&viewFallbackFrom=aspnetcore-3.0&pivots=webassembly#request-a-service-in-a-component).

This is one of those cases where the lightweight abstraction that ships with the framework (Blazor) does a good job of shielding the underlying implementation, and makes it easy enough to swap out at a later date.

From your exploration @harperjohn Autofac does sound good as an option 'down the line' if we run into any limitations with service lifetimes etc.

**Comment by harperjohn** [comment]
One nagging/recurring thought that keeps coming to mind with DI - is "memory leaks" and "memory leak detection"

Does anyone have experience with this? In any environment, these can be the most troublesome bugs.

**Comment by jonhilt** [comment]
Good question @harperjohn! 

Can only speak from my own experience but the last five or six projects I worked on all used DI (specifically MS DI via .NET core), vertical slice architecture (using MediatR) and EF Core (in various versions) and we didn't run into any significant performance issues or memory leaks.

We also completely swapped out data persistence in one of the projects from an event-sourced approach using Postgres and Marten to SQL and EF/Dapper with little to zero application downtime, but that's another story for another day :-)

**Comment by AnkurSheel** [comment]
> One nagging/recurring thought that keeps coming to mind with DI - is "memory leaks" and "memory leak detection"

Echoing what Jon said but from experience in previous projects with decent daily usage - Used MS DI via .Net Core. Everything a singleton (because of some other constraints). Haven't run into performance issues so far.
